feat: Implement Task Sharing Feature

ShubhamAggarwal0812 opened the pull request:
# Pull Request for Task Sharing Feature

## Description
This PR introduces the task-sharing feature to our application, allowing users to share tasks with other users. Additionally, it updates the existing code to fetch accounts to display on the shared task modal. The implementation includes both backend and frontend changes to support this functionality.

## Why is this change needed?
This change is necessary to enable users to share tasks with other users, facilitating collaboration and task management within teams. The feature enhances the application by providing users with the ability to delegate and track shared tasks.

## Changes Overview
### Backend
1. **New Endpoints:**
   - `POST /shared-tasks`: Share a task with one or more users.
   - `GET /shared-tasks`: Fetch all tasks shared with the current user.

2. **Database Models:**
   - **SharedTaskDB Schema:** Defines the structure of shared tasks stored in the database.

3. **Controllers:**
   - **SharedTaskController:** Handles the HTTP requests for shared tasks.

4. **Services:**
   - **SharedTaskService:** Contains the business logic for managing shared tasks.

5. **Utils:**
   - **SharedTaskUtil:** Converts database objects to API responses.

6. **Updates:**
   - Updated existing code to fetch accounts for displaying in the shared task modal.

### Frontend
1. **Components:**
   - **SharedTaskList:** Displays a list of tasks shared with the user.
   - **ShareTaskModal:** Allows users to share a task with other users.

2. **Context:**
   - **SharedTaskContext:** Manages the state and actions related to shared tasks.

3. **Services:**
   - **SharedTaskService:** Handles API calls related to shared tasks.

4. **Types:**
   - **SharedTask:** Defines the structure of shared task objects used in the frontend.

## API Changes
- **POST /shared-tasks**
  - **Request Body:**
    ```json
    {
      "taskId": "string",
      "accountIds": ["string"]
    }
    ```
  - **Response:**
    ```json
    [
      {
        "id": "string",
        "task": {
          "id": "string",
          "title": "string",
          "description": "string",
          "account": {
            "id": "string",
            "firstName": "string",
            "lastName": "string",
            "username": "string"
          }
        },
        "account": "string"
      }
    ]
    ```

- **GET /shared-tasks**
  - **Response:**
    ```json
    [
      {
        "id": "string",
        "task": {
          "id": "string",
          "title": "string",
          "description": "string",
          "account": {
            "id": "string",
            "firstName": "string",
            "lastName": "string",
            "username": "string"
          }
        },
        "account": "string"
      }
    ]
    ```

## Database Schema Changes
No changes were made to the existing database schema. The new `SharedTaskDB` schema was added to support task sharing.

## Tests
### Automated Test Cases Added: (Yet to Add)
1. **Creating a Shared Task:** Validates that a task can be shared successfully.
2. **Fetching Shared Tasks:** Ensures that tasks shared with the user are fetched correctly.

### Manual Test Cases Run:
1. **Creating a Shared Task:**
   - Navigate to a task.
   - Share the task with other users.
   - Verify that the shared task appears in the shared task list.
2. **Fetching Shared Tasks:**
   - Refresh the shared tasks page.
   - Ensure that all shared tasks are displayed correctly.

## Meta Information
- **Assignee:** @ShubhamAggarwal0812 
- **Labels:** feature, task-sharing, backend, frontend

## Self Review
This PR has been self-reviewed to ensure quality and consistency.

Thank you for reviewing this PR.

Commit:
feat: Implement Task Sharing Feature

aniketrathi commented on src/apps/backend/app.ts:
```diff
@@ -104,9 +105,7 @@ export default class App {
-      transports: [
-        new CustomLoggerTransport(),
-      ],
+      transports: [new CustomLoggerTransport()],
```
There are no actual changes, primarily whitespace/formatting-related changes. We should ensure that the VS editor uses the prettier config checked in the code base to avoid this

aniketrathi commented on src/apps/backend/modules/account/account-service.ts:
```diff
@@ -92,10 +96,12 @@ export default class AccountService {
-      lastName,
-    );
+    return AccountWriter.updateAccountDetails(accountId, firstName, lastName);
+  }
```
Same there are no actual changes except the `getAllAccounts` method.

aniketrathi commented on src/apps/backend/modules/account/rest-api/account-controller.ts:
```diff
@@ -19,15 +20,9 @@ export class AccountController {
-      } = req.body as CreateAccountParamsByPhoneNumber;
+      const { firstName, lastName, password, username } =
+        req.body as CreateAccountParamsByUsernameAndPassword;
+      const { phoneNumber } = req.body as CreateAccountParamsByPhoneNumber;
```
Same there are no actual changes

aniketrathi commented on src/apps/backend/modules/shared-task/internal/shared-task-util.ts:
```diff
@@ -0,0 +1,14 @@
+  ): SharedTask {
+    const sharedTask = new SharedTask();
+    sharedTask.id = sharedTaskDb._id.toString();
+    (sharedTask.task = sharedTaskDb.task),
```
remove comma and parentheses.
`sharedTask.task = sharedTaskDb.task` is not correct. We have to do it this way if the type of sharedTaskDb.task is object id then `sharedTask.task = sharedTaskDb.task.toString()` else if it is an object then `sharedTask.task = TaskUtil.convertTaskDBToTask(sharedTaskDb.task)`.

aniketrathi commented on src/apps/backend/modules/shared-task/types.ts:
```diff
@@ -0,0 +1,37 @@
+
+export class SharedTask {
+  id: string;
+  task: any;
```
Avoid the use of type any. Use specific types for better safety.
task: string | Task;
account: string | Account;

aniketrathi commented on src/apps/frontend/app.component.tsx:
```diff
@@ -21,10 +22,14 @@ export default function App(): React.ReactElement {
-          </Router>
+          <SharedTaskProvider>
+            {' '}
+            {/* Add this provider */}
```
What is this for?
 {' '}
 {/* Add this provider */}

aniketrathi commented on src/apps/backend/modules/shared-task/internal/store/shared-task-db.ts:
```diff
@@ -0,0 +1,31 @@
+      ref: 'Account',
+      index: true,
+      required: true,
+    },
```
we can add an active field here so that we can support the unshare of tasks functionality in the future.

Commit:
feat: Refactor shared-task module, util, types, and modal

ShubhamAggarwal0812 commented:
### PR Update: Implementations Based on Feedback

I have refactored both the backend and frontend code for the shared task feature as per the feedback provided. Below are the changes made:

#### Backend
1. **Shared Task Util:**
   - Refactored `convertSharedTaskDBToSharedTask` to handle both `ObjectId` and `Task` types for `sharedTask.task`.
   - Updated to properly convert `task` and `account` fields using appropriate utility methods.

2. **Shared Task Types:**
   - Replaced the use of `any` with specific types for `sharedTask.task` and `sharedTask.account`.
   - Added proper type handling for `task` and `account` fields.

3. **Shared Task DB Schema:**
   - Added an `active` field to the `SharedTaskDB` schema to support soft deletion of shared tasks.

4. **Soft Deletion Handling:**
   - Implemented logic to remove shared tasks when the associated task is deleted by setting the `active` field to `false`.

5. **Shared Task Serializer:**
   - Modified the serializer to handle both `string` and `Task` types for `sharedTask.task`.

#### Frontend
1. **Shared Task Modal:**
   - Enhanced `ShareTaskModal` to display accounts with proper spacing and formatting.

These changes ensure proper integration between the frontend and backend while adhering to the feedback provided. The code has been tested to confirm that it works as expected.

Please review the updates and let me know if there are any further changes required.

aniketrathi commented on src/apps/backend/modules/shared-task/internal/shared-task-util.ts:
```diff
@@ -0,0 +1,55 @@
+    if (Types.ObjectId.isValid(sharedTaskDb.task.toString())) {
+      sharedTask.task = sharedTaskDb.task.toString();
+    } else {
+      sharedTask.task = SharedTaskUtil.convertTask(sharedTaskDb.task);
```
If you want `Task` after all, why are you not using `convertTaskDBToTask` from `TaskUtil`?

If you want something different from `Task` for this populated task then you can create a method like you have created `convertTask`.

aniketrathi commented on src/apps/backend/modules/shared-task/internal/shared-task-util.ts:
```diff
@@ -0,0 +1,55 @@
+      sharedTask.task = SharedTaskUtil.convertTask(sharedTaskDb.task);
+    }
+
+    sharedTask.account = SharedTaskUtil.convertAccount(sharedTaskDb.account);
```
Same goes with this

aniketrathi commented:
Please write automated test cases.
